networkx/algorithms/decomposition.py

```diff
@@ -0,0 +1,174 @@
+"""
+Find the decompositions of a graph.
+
+Two techniques are implemented: k-core and k-peak.
+
+The highest possible k (for core or peak) is called degeneracy.
+
+See the following references for details:
+
+Graph core: https://networkx.org/documentation/stable/reference/algorithms/core.html
+
+The k-peak Decomposition: Mapping the Global Structure of Graphs, Govindan Priya, Wang Chenghong, Xu Chumeng, Duan Hongyu, and Soundarajan Sucheta, WWW '17. https://doi.org/10.1145/3038912.3052635
+
+"""
+
+import networkx as nx
+
+__all__ = [
+    "k_core_decomposition",
+    "k_core_degeneracy",
+    "k_peak_decomposition",
+    "k_peak_degeneracy",
+]
+
+
+@nx.utils.not_implemented_for("multigraph")
+def k_core_decomposition(G, k, core_number=None):
+    """Returns the k-core decomposition of G.
+
+    The k-core decomposition is the subgraph of G that each node has `k` or more degree.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+      A graph or directed graph
+    k : int
+      Value of the decomposition
+    core_number : dictionary, optional
+      Precomputed core numbers for the graph G.
+
+    Returns
+    -------
+    G : NetworkX graph
+      The k-core decomposition
+
+    Raises
+    ------
+    NetworkXNotImplemented
+      The k-core decomposition is not defined for multigraphs or graphs with self loops.
+
+    See Also
+    --------
+    core_number
+
+    """
+    return nx.k_core(G, k=k, core_number=core_number)
+
+
+@nx.utils.not_implemented_for("multigraph")
+def k_core_degeneracy(G, core_number=None):
+    """Returns the k-core degeneracy of G.
+
+    The k-core degeneracy is the subgraph of G that each node has maximum core number degree.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+      A graph or directed graph
+    core_number : dictionary, optional
+      Precomputed core numbers for the graph G.
+
+    Returns
+    -------
+    G : NetworkX graph
+      The k-core degeneracy
+
+    Raises
+    ------
+    NetworkXNotImplemented
+      The k-core degeneracy is not defined for multigraphs or graphs with self loops.
+
+    Notes
+    -----
+    K-core degeneracy is the highest k-core decomposition.
+
+    See Also
+    --------
+    core_number
+    k_core_decomposition
+
+    """
+    if core_number is None:
+        core_number = nx.core_number(G)
+    k = max(core_number.values())
+
+    return nx.k_core(G, k=k, core_number=core_number)
```

Comment:
This feels like a good example for the docstring of `k_core` rather than adding a whole new function.

Reply:
I actually have a large swath of decompositions incoming for large model modifications and analysis. While this project may not augment with them, we have an internal project you may be able to contribute Ahmet Tüzen. Ping me here, and folks - good work on this project. Cheers.